indico/modules/events/models/persons.py

```diff
@@ -146,6 +147,65 @@ def for_user(cls, user, event):
         person = event.persons.filter_by(user=user).first()
         return person or cls.create_from_user(user, event)
 
+    @classmethod
+    def replace_email_with_user(cls, user, email_address=None):
+        """
+        Replaces all email-based persons matching the user's email
+        addresses with user-based entries.
+
+        :param user: A User object.
+        :param email_address: Email address
+        """
+        from indico.modules.events.models.events import Event
+        query = (cls.query
+                 .options(noload('*'))
+                 .join(EventPerson.event_new)
+                 .filter(~Event.is_deleted, cls.email == email_address, cls.user_id.is_(None)))
+        for event_person in query:
+            existing = (cls.query
+                        .options(noload('*'))
+                        .filter_by(user=user, event_id=event_person.event_id)
+                        .join(EventPerson.user)
+                        .first())
+            if existing is None:
+                event_person.user = user
+            else:
+                existing.merge_person_info(event_person)
+                db.session.delete(event_person)
+        db.session.flush()
+
+    def merge_person_info(self, other):
+        from indico.modules.events.contributions.models.persons import AuthorType
+        for column_name in {'_title', 'affiliation', 'address', 'phone', 'first_name', 'last_name'}:
+            value = getattr(self, column_name) or getattr(other, column_name)
+            setattr(self, column_name, value)
+
+        def _get_author_type(src_link, dest_link):
+            if AuthorType.primary in (src_link.author_type, dest_link.author_type):
+                return AuthorType.primary
+            if src_link.author_type == AuthorType.none:
+                return dest_link.author_type
+
+        for event_link in other.event_links:
+            if self.event_links.filter_by(event_id=event_link.event_id).first():
+                db.session.delete(event_link)
+            else:
+                event_link.person = self
+        for contribution_link in other.contribution_links:
+            contrib_link = self.contribution_links.filter_by(contribution_id=contribution_link.contribution_id).first()
+            if contrib_link:
+                contrib_link.is_speaker |= contribution_link.is_speaker
+                contrib_link.author_type = _get_author_type(contrib_link, contribution_link)
```

Comment:
`contrib_link` and `contribution_link`.... please find another name for one of them, otherwise it's extremely confusing